### PR TITLE
Add new user_irq claim APIs to make it easier for independent code using them to interoperate

### PR DESCRIPTION
**Pull request by kilograham**

### CODE REVIEW

File: src/rp2_common/hardware_irq/include/hardware/irq.h [hunk 1]
```diff
@@ -289,6 +289,65 @@ void irq_set_pending(uint num);
  * \note This is an internal method and user should generally not call it.
  */
 void irq_init_priorities(void);
+
+/*! \brief Claim ownership of a user IRQ on the calling core
+ *  \ingroup hardware_irq
+ *  
+ * User IRQs are numbered 26-31 and are not connected to any hardware, but can be triggered by \ref irq_set_pending.
+ *
+ * \note User IRQs are entirely core local.
```
Comment (lurch) [inline src/rp2_common/hardware_irq/include/hardware/irq.h:298]: Does "entirely core local" mean that each core has it's own independent set of User IRQs? Might be worth being slightly more verbose? :shrug:

File: src/rp2_common/hardware_irq/irq.c [hunk 2]
```diff
@@ -410,3 +413,31 @@ void irq_init_priorities() {
...
+    hw_claim_or_assert(&user_irq_claimed[get_core_num()], get_user_irq_claim_index(irq_num), "User IRQ is already claimed");
+}
+
+void user_irq_unclaim(uint irq_num) {
+    hw_claim_clear(&user_irq_claimed[get_core_num()], get_user_irq_claim_index(irq_num));
+}
+
+int user_irq_claim_unused(bool required) {
+    int bit = hw_claim_unused_from_range(&user_irq_claimed[get_core_num()], required, 0, NUM_USER_IRQS-1, "No user IRQs are available");
+    if (bit >= 0) bit =  (int)NUM_IRQS - bit -1;
```
Comment (lurch) [inline src/rp2_common/hardware_irq/irq.c:436]: Is it worth reusing `get_user_irq_claim_index` here?
(also, weird spacing here, with double space after `=` and no space between `-` and `1`)

File: src/rp2_common/hardware_irq/irq.c [hunk 3]
```diff
@@ -410,3 +413,31 @@ void irq_init_priorities() {
     }
 #endif
 }
+
+#define FIRST_USER_IRQ (NUM_IRQS - NUM_USER_IRQS)
+
+static uint get_user_irq_claim_index(uint irq_num) {
+    invalid_params_if(IRQ, irq_num < FIRST_USER_IRQ || irq_num >= NUM_IRQS);
+    // we count backwards from the last, to match the existing hard coded uses of user IRQs in the SDK which were previously using 31
+    static_assert(NUM_IRQS - FIRST_USER_IRQ - 1 <= 8, ""); // we only use a single byte's worth of claim bits today.
```
Comment (Wren6991) [inline src/rp2_common/hardware_irq/irq.c:422]: This looks to be off by one.

If number of IRQs is 32, and first user IRQ is 31, there is 1 user IRQ, but this expression would give 0.

**Review by Wren6991 — COMMENTED** [review]
Happy with this, except for the potential off-by-one in the static assert (currently harmless) and @lurch's comments.

File: src/rp2_common/pico_stdio_usb/include/pico/stdio_usb.h [hunk 4]
```diff
@@ -39,9 +39,9 @@
 #define PICO_STDIO_USB_TASK_INTERVAL_US 1000
 #endif
 
-// PICO_CONFIG: PICO_STDIO_USB_LOW_PRIORITY_IRQ, low priority (non hardware) IRQ number to claim for tud_task() background execution, default=31, advanced=true, group=pico_stdio_usb
+// PICO_CONFIG: PICO_STDIO_USB_LOW_PRIORITY_IRQ, low priority (non hardware) IRQ number to claim for tud_task() background execution, advanced=true, group=pico_stdio_usb
```
[the remark this answers is not in the document]
Comment (kilograham) [reply src/rp2_common/pico_stdio_usb/include/pico/stdio_usb.h:42]: I think i shall leave it for now, but i will change the description to make it less attractive

**Pull request merged**